[Models] Add remaining model PP support

andoorve opened the pull request:
Add all other possible models for PP

FIX #7684
FIX #8923

Commit:
Model PP support

Signed-off-by: Muralidhar Andoorveedu <muralidhar.andoorveedu@centml.ai>

youkaichao commented:
Thanks for the great effort! Please lmk when it is ready for review.

andoorve commented:
I think you can already start taking a look @youkaichao with @ywang96. I think the major thing here is to ensure all the models work after this change.

Commit:
Format

Signed-off-by: Muralidhar Andoorveedu <muralidhar.andoorveedu@centml.ai>

Commit:
Format

Signed-off-by: Muralidhar Andoorveedu <muralidhar.andoorveedu@centml.ai>

Commit:
Merge branch 'main' of github.com:andoorve/vllm into qwen-pp

Commit:
Format

Signed-off-by: Muralidhar Andoorveedu <muralidhar.andoorveedu@centml.ai>

andoorve commented:
Hey @ywang96 can you take a quick look at vision model stuff?

Commit:
Merge branch 'main' of github.com:andoorve/vllm into qwen-pp

Signed-off-by: Murali Andoorveedu <muralidhar.andoorveedu@centml.ai>

Commit:
Merge branch 'main' of github.com:andoorve/vllm into qwen-pp

Signed-off-by: Murali Andoorveedu <muralidhar.andoorveedu@centml.ai>

Commit:
Format

Signed-off-by: Murali Andoorveedu <muralidhar.andoorveedu@centml.ai>

andoorve commented:
@ywang96 Should be rebased for you - it's untested though!

Commit:
Merge branch 'main' of github.com:andoorve/vllm into qwen-pp

DarkLight1337 commented:
@ywang96 is busy this week so I'll be taking over the review. I'll add each model to PP test and test them locally.

DarkLight1337 committed:
Merge branch 'main' into qwen-pp

DarkLight1337 committed:
fix wrong type

DarkLight1337 committed:
fix typo

DarkLight1337 committed:
Merge branch 'main' into qwen-pp

DarkLight1337 committed:
Add SupportsPP interface and stateless protocol check

DarkLight1337 committed:
Subclass SupportsPP in relevant models

DarkLight1337 committed:
Remove hardcoded list

DarkLight1337 committed:
Remove unused import

DarkLight1337 committed:
Check using function

DarkLight1337 committed:
Update docstring

DarkLight1337 committed:
Simplify

DarkLight1337 committed:
Add tests

DarkLight1337 committed:
Test CUDA initialization

DarkLight1337 committed:
Add platform guard

DarkLight1337 committed:
Trigger CI

DarkLight1337 committed:
Fix OOT registration

DarkLight1337 committed:
Update docstring

DarkLight1337 committed:
Remove unnecessary global

DarkLight1337 committed:
Merge branch 'main' into qwen-pp

DarkLight1337 committed:
Update interfaces

DarkLight1337 committed:
format

DarkLight1337 committed:
Merge branch 'supports-pp' into qwen-pp

DarkLight1337 committed:
Inherit from `SupportsPP`

DarkLight1337 committed:
Merge branch 'main' into qwen-pp

DarkLight1337 committed:
Inherit from `SupportsPP`

DarkLight1337 committed:
Fix error check

DarkLight1337 committed:
Make `prefix` required

DarkLight1337 committed:
Merge branch 'supports-pp' into qwen-pp

DarkLight1337 commented:
@andoorve I'm running the tests locally and found a few failing models so I'm debugging them now. Before I update this PR, can you take a look at #9000? It simplifies the process of adding new PP models.

DarkLight1337 committed:
Fix PP for language models

DarkLight1337 committed:
Fix environment variables not being copied over

DarkLight1337 committed:
Merge branch 'main' into supports-pp

DarkLight1337 committed:
Merge branch 'supports-pp' into qwen-pp

DarkLight1337 committed:
Use inferred type

DarkLight1337 committed:
Add missing type annotations

DarkLight1337 committed:
Add PP support for more multimodal models
- Remove unnecessary `is_pp_missing_parameter`
- Fix missing `sampler` attribute
- Fix multimodal inputs being embedded multiple times in PP
- Updated `_MULTIMODAL_MODEL_MAX_NUM_BATCHED_TOKENS` to be sufficient for llava-next-video

DarkLight1337 committed:
Fix the real problem, which is that modelscope is not installed

DarkLight1337 committed:
Merge branch 'supports-pp' into qwen-pp

DarkLight1337 committed:
Update tests

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Fix missing `SupportsPP`; support PP for olmoe

DarkLight1337 committed:
Fix type annotations

DarkLight1337 committed:
Move modelscope installation into regression test

DarkLight1337 committed:
Merge branch 'supports-pp' into qwen-pp

DarkLight1337 committed:
format

DarkLight1337 committed:
Update LoRA support in docs

DarkLight1337 committed:
PP support for phimoe

DarkLight1337 committed:
Fix capitalization

DarkLight1337 commented:
Most models work with PP now. My machine doesn't have enough memory to test the following models though:
- Arctic
- Command-R
- DBRX
- Mixtral
- Solar Pro

The following models support PP but are unable to be tested at all:
- Phi-3.5-MoE (cannot be used in online serving, see #8553)
- XVERSE (the model cannot be used in the current version of transformers/vLLM at all because of outdated tokenizer config)

The following models still don't support PP:
- Jamba
- BART
- Llama 3.2

Finally, the following models support PP but cannot be tested:

DarkLight1337 committed:
Fix `LLMWrapper`

DarkLight1337 committed:
Merge branch 'supports-pp' into qwen-pp

DarkLight1337 committed:
Update test configs

youkaichao commented on tests/distributed/test_pipeline_parallel.py:
do we need to test so many models? it might blow up the ci

DarkLight1337 replied:
I only test a few models in the CI. The rest are just definitions of test settings so it's easy to test individual models locally.

andoorve replied:
Just wondering should we keep it directly in the test file, or is it good to add this to some kind of developer FAQ?

youkaichao commented on vllm/model_executor/models/__init__.py:
```diff
@@ -200,21 +225,99 @@ def register_model(model_arch: str, model_cls: Type[nn.Module]):
+            f"assert {func.__name__}({cls_name}), '{err_id}'",
+        ])
+
+        result = subprocess.run([sys.executable, "-c", stmts],
```
how many times will it run, when we serve a model?

creating a subprocess can be expensive.

DarkLight1337 replied:
Since the result is cached, it will only be run at most `2 * len(architectures)` per model (if both pp and multimodal are checked).

DarkLight1337 replied:
The startup time should be short compared to the time it takes to load the model weights.

youkaichao reviewed (approved):
thanks for the amazing effort!

andoorve commented:
@DarkLight1337 Thank you for your help! For #9000, can we pull all the changes into this PR and push it together?

DarkLight1337 commented:
> @DarkLight1337 Thank you for your help! For #9000, can we pull all the changes into this PR and push it together?

This PR already includes all the changes from that PR.

andoorve commented:
I see you mentioned this:

> We should merge https://github.com/vllm-project/vllm/pull/9000 first though to reduce the diffs.

Are the diffs resolved in this PR? I.e. can we merge this directly and delete #9000?

DarkLight1337 commented:
> I see you mentioned this:
> 
> > We should merge #9000 first though to reduce the diffs.
> 
> Are the diffs resolved in this PR? I.e. can we merge this directly and delete #9000?

Yes.

andoorve commented:
Need to ungate Pixtral for testing: https://huggingface.co/mistralai/Pixtral-12B-2409

andoorve commented:
@youkaichao @DarkLight1337 Possible to add this model to the HF account of the testing token?

youkaichao commented:
> @youkaichao @DarkLight1337 Possible to add this model to the HF account of the testing token?

added by @simon-mo

youkaichao commented:
it looks pixtral causes oom

andoorve commented:
Hmm how is that possible? Pixtral is only 12B and we have 4 L4

andoorve commented:
Also tested DBRX and Arctic locally and they passed. Didn't add to test configs cause I had to modify them to work.

andoorve committed:
Add more tp to pixtral

Pixtral can't fit on single GPU

andoorve committed:
Update test_pipeline_parallel.py

Add noqa

andoorve commented:
Decoder only LM Models test failure:
```

[2024-10-03T21:16:08Z] FAILED models/decoder_only/language/test_big_models.py::test_models[32-half-EleutherAI/gpt-j-6b] - KeyError: 'transformer.h.9.ln_1.weight'
--
  | [2024-10-03T21:16:08Z] FAILED models/decoder_only/language/test_big_models.py::test_model_print[half-EleutherAI/gpt-j-6b] - KeyError: 'transformer.h.9.ln_1.weight'
  | [2024-10-03T21:16:08Z] FAILED models/decoder_only/language/test_models.py::test_models[96-float-facebook/opt-125m] - RuntimeError: value cannot be converted to type at::Half without overflow
```

andoorve committed:
Fix gpt_j.py

andoorve commented:
Fixed error 2&3, error 1 passed when testing locally

andoorve commented:
Looks like fix to pixtral worked.

andoorve commented:
@youkaichao @DarkLight1337 tests pass fully

DarkLight1337 commented:
Thanks for your hard work as well!

youkaichao commented:
@andoorve @DarkLight1337 thanks for the great work!!!

sekh77 commented:
@youkaichao - Is this change now available in version 0.6.2? I have a requirement to load LLaMA 3.12 90B model across four GPUs spread across two nodes using pipeline parallel.

youkaichao commented:
> @youkaichao - Is this change now available in version 0.6.2? I have a requirement to load LLaMA 3.2 90B vision model across four GPUs spread across two nodes using pipeline parallel.

not yet, but you can always use the latest wheel from the main branch. see https://docs.vllm.ai/en/latest/getting_started/installation.html#install-the-latest-code

eByteTheDust commented:
> > @youkaichao - Is this change now available in version 0.6.2? I have a requirement to load LLaMA 3.2 90B vision model across four GPUs spread across two nodes using pipeline parallel.
> 
> not yet, but you can always use the latest wheel from the main branch. see https://docs.vllm.ai/en/latest/getting_started/installation.html#install-the-latest-code

Can you please confirm if the latest wheel supports PP for llama 3.2 11B Instruct?

DarkLight1337 commented:
> > > @youkaichao - Is this change now available in version 0.6.2? I have a requirement to load LLaMA 3.2 90B vision model across four GPUs spread across two nodes using pipeline parallel.
> > 
> > 
> > not yet, but you can always use the latest wheel from the main branch. see https://docs.vllm.ai/en/latest/getting_started/installation.html#install-the-latest-code
> 
> Can you please confirm if the latest wheel supports PP for llama 3.2 11B Instruct?

Llama-3.2-Vision specifically isn't PP-supported because PP isn't supported in general for encoder-decoder models yet.

sekh77 commented:
@DarkLight1337 -  Is PP supported for DataBricks DBRX model - databricks/dbrx-instruct?

DarkLight1337 commented:
Yes. Please check out the [Supported Models page](https://docs.vllm.ai/en/latest/models/supported_models.html) for more details.

aaron-asdf commented:
Is there a PR to follow to know about the status of vision based encoder-decoder models?
